src/Simplex/Messaging/Agent/NtfSubSupervisor.hs

```diff
@@ -0,0 +1,117 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE NamedFieldPuns #-}
+
+module Simplex.Messaging.Agent.NtfSubSupervisor
+  ( NtfSubSupervisor (..),
+    RcvQueueNtfCommand (..),
+    newNtfSubSupervisor,
+    addNtfSubWorker,
+    addNtfSubSMPWorker,
+    removeNtfSubWorker,
+    removeNtfSubSMPWorker,
+    nsUpdateToken,
+    nsUpdateToken',
+    nsRemoveNtfToken,
+    addRcvQueueToNtfSubQueue,
+    addRcvQueueToNtfSubQueue',
+    closeNtfSubSupervisor,
+  )
+where
+
+import Control.Concurrent.Async (Async, uninterruptibleCancel)
+import Control.Monad
+import Control.Monad.IO.Unlift (MonadUnliftIO)
+import Simplex.Messaging.Agent.Env.SQLite
+import Simplex.Messaging.Agent.Monad (AgentMonad)
+import Simplex.Messaging.Agent.Store
+import Simplex.Messaging.Client.Agent ()
+import Simplex.Messaging.Notifications.Client
+import Simplex.Messaging.Notifications.Protocol (NtfTknStatus (NTActive))
+import Simplex.Messaging.Protocol (ProtocolServer (..), SMPServer)
+import Simplex.Messaging.TMap (TMap)
+import qualified Simplex.Messaging.TMap as TM
+import UnliftIO (async)
+import UnliftIO.STM
+
+data NtfSubSupervisor = NtfSubSupervisor
+  { ntfTkn :: TVar (Maybe NtfToken),
+    ntfSubQ :: TBQueue (RcvQueue, RcvQueueNtfCommand),
+    ntfSubWorkers :: TMap NtfServer (TMVar (), Async ()),
+    ntfSubSMPWorkers :: TMap SMPServer (TMVar (), Async ())
+  }
+
+data RcvQueueNtfCommand = RQNCCreate | RQNCDelete
+
+newNtfSubSupervisor :: Env -> STM NtfSubSupervisor
+newNtfSubSupervisor agentEnv = do
+  ntfTkn <- newTVar Nothing
+  ntfSubQ <- newTBQueue (ntfSubTbqSize $ config agentEnv)
+  ntfSubWorkers <- TM.empty
+  ntfSubSMPWorkers <- TM.empty
+  pure
+    NtfSubSupervisor
+      { ntfTkn,
+        ntfSubQ,
+        ntfSubWorkers,
+        ntfSubSMPWorkers
+      }
+
+addNtfSubWorker :: MonadUnliftIO m => NtfSubSupervisor -> NtfServer -> (TMVar () -> m ()) -> m ()
+addNtfSubWorker ns = addNtfSubWorker_ $ ntfSubWorkers ns
+
+addNtfSubSMPWorker :: MonadUnliftIO m => NtfSubSupervisor -> SMPServer -> (TMVar () -> m ()) -> m ()
+addNtfSubSMPWorker ns = addNtfSubWorker_ $ ntfSubSMPWorkers ns
+
+addNtfSubWorker_ :: MonadUnliftIO m => TMap NtfServer (TMVar (), Async ()) -> ProtocolServer -> (TMVar () -> m ()) -> m ()
+addNtfSubWorker_ workerMap srv action =
+  atomically (TM.lookup srv workerMap) >>= \case
+    Nothing -> do
+      workerSemaphore <- newTMVarIO ()
```

Comment:
"workerSemaphore" - this name should describe what it is semaphoring, not that it is some semaphore...